pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+	defer klog.Infof("Shutting down Multi CIDR Range allocator")
+
+	if !cache.WaitForNamedCacheSync("multi_cidr_range_allocator", stopCh, r.nodesSynced, r.clusterCIDRConfigSynced) {
+		return
+	}
+
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		go r.worker(stopCh)
+		go r.cccWorker()
+	}
+
+	<-stopCh
+}
+
+func (r *multiCIDRRangeAllocator) worker(stopChan <-chan struct{}) {
+	for {
+		select {
+		case workItem, ok := <-r.nodeCIDRUpdateChannel:
+			if !ok {
+				klog.Warning("Channel nodeCIDRUpdateChannel was unexpectedly closed")
+				return
+			}
+			if err := r.updateCIDRsAllocation(workItem); err != nil {
+				// Requeue the failed node for update again.
+				r.nodeCIDRUpdateChannel <- workItem
+			}
+		case <-stopChan:
+			return
+		}
+	}
+}
+
+func (r *multiCIDRRangeAllocator) cccWorker() {
+	for r.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (r *multiCIDRRangeAllocator) processNextWorkItem() bool {
+	cKey, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(cKey)
+
+	err := r.syncClusterCIDRConfig(cKey.(string))
+	r.handleErr(err, cKey)
+
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) handleErr(err error, key interface{}) {
+	if err == nil {
+		r.queue.Forget(key)
+		return
+	}
+
+	if r.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error syncing clusterCIDRConfig %q, retrying. Error: %v", key, err)
+		r.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping clusterCIDRConfig %q out of the queue: %v", key, err)
+	r.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (r *multiCIDRRangeAllocator) syncClusterCIDRConfig(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing clusterCIDRConfig request %q (%v)", key, time.Since(startTime))
+	}()
+
+	clusterCIDRConfig, err := r.clusterCIDRConfigLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		klog.V(3).Infof("clusterCIDRConfig has been deleted: %v", key)
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check the DeletionTimestamp to determine if object is under deletion.
+	if !clusterCIDRConfig.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(clusterCIDRConfig)
+	}
+	return r.reconcileCreate(clusterCIDRConfig)
+}
+
+func (r *multiCIDRRangeAllocator) insertNodeToProcessing(nodeName string) bool {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	if r.nodesInProcessing.Has(nodeName) {
+		return false
+	}
+	r.nodesInProcessing.Insert(nodeName)
+	return true
+}
+
+func (r *multiCIDRRangeAllocator) removeNodeFromProcessing(nodeName string) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	klog.Infof("Removing Node: %s from processing", nodeName)
+	r.nodesInProcessing.Delete(nodeName)
+}
+
+// marks node.PodCIDRs[...] as used in allocator's tracked cidrSet.
+func (r *multiCIDRRangeAllocator) occupyCIDRs(node *v1.Node) error {
+	defer r.removeNodeFromProcessing(node.Name)
```

Comment:
don't use defer like this. It makes it very hard to verify the code paths.
if you always want to do this on all codepaths, you can nest func

occupyCIDRs() {
  occupyCIDRsImpl()
  r.removeNodeFromProcessing()
}

Reply:
Also, we might want to consider call ing this directly while holding the lock below

Reply:
Done